### ISSUE TITLE
Possible ip clash with gateway?

### ISSUE THREAD

**Issue by Raven888888** [flannel-io/flannel#1707]
*I believe this is not an issue for flannel per se, correct me if I am wrong. I am quite new to networking. ;)*

**Context**
I am trying to setup Kubernetes cluster in the Dedicated bare metal servers, with the 3 servers (call them Kube 1, Kube 2 and Kube Storage) and Flannel CNI. By default, Kubernetes has its own internal service discovery CoreDNS with IP 10.96.0.10, which is obtained from CIDR range 10.96.0.0/12. 

Kube 1 and Kube 2 are on the same physical rack and using the same network switch. Kube Storage is on a different rack.

I limit the CoreDNS pod to run only in Kube 1. The problem I found is inconsistent dns lookup from Kube Storage to the pod, mostly just fail to reach CoreDNS. But dns lookup from Kube 1 and Kube 2 to the pod is very quick and consistent.

I suspect it is due to my service provider also uses 10.x.x.x as the gateway switches. Because when I did traceroute, I can see that.

Why from Kube 1 and Kube 2 to the CoreDNS pod (10.96.0.10) does not have the issue? Because these servers are on the same physical rack, and therefore shares the same switch. It manages to reach CoreDNS pod without going through many switches beyond the rack.

But for Kube Storage to CoreDNS, because they are on different racks, it has to go through many switches which has IP clash with CoreDNS, and it mostly never able to reach the pod. Occasionally, 1 out 10 times, it can reach CoreDNS pod and resolve fine, but it is slow.

To further confirm this finding, I change the CoreDNS pod to run in Kube Storage instead, at the same service IP 10.96.0.10. Same observations, but this time inverted:
- From Kube Storage to CoreDNS pod, dns lookup is fast and consistent.
- From Kube 1 or Kube 2 to CoreDNS pod, dns lookup mostly fails to reach CoreDNS pod.

To further confirm this finding again, I set up private network among the 3 servers, with private IPs. CoreDNS is still running at 10.96.0.10.
Now, all 3 servers are able to reach CoreDNS pod, and dns lookup are quick. And the explanation is private network does not go through all the gateway switches, hence no IP clash.

*I believe Flannel is doing what it should do, no issues with that. I believe the problem is IP clash.*

So the question is:
- Is my hypothesis correct?
- If yes, how can I change the cluster CIDR and service ip range in Flannel, so that I do not need to resetup the cluster?

**Comment by rbrtbnfgl** [comment]
If you are using the default config as backend the traffic to the pod should be forwarded using the VXLAN tunnel. On the network side the gateway forwards UDP VXLAN packets with the IP of the nodes. If there are any overlap between the PODs addresses and the network you could get some issues with the forwarded packets.
On kubernetes you can't change the allocated addresses on a running cluster.

**Comment by Raven888888** [comment]
Thank you @rbrtbnfgl 

Yes I am using UDP VXLAN in my setup.

>  On kubernetes you can't change the allocated addresses on a running cluster.  

So I have to resetup the whole cluster with a different clusterCIDR and service ip range, in order to avoid the issue? There is no way to modify it in the current cluster? (Because I already have some loads running in it)

**Comment by rbrtbnfgl** [comment]
The right way to do it should be doing a backup of the running config and pods and restart the cluster with the new CIDR.

**Comment by Raven888888** [comment]
Thanks @rbrtbnfgl 

But I think my issue is not with CIDR, but with the interface that flannel is using. Because I tried resetup cluster with 2 different ranges of CIDR (11.244.0.0/16 & 160.244.0.0/16) and service ranges (11.96.0.0/12 & 160.96.0.0/12), I still see the issue of intermittent dns resolution issues.

Of all the setup, I was using public LAN as the interface in each flannel pod. UDP packets get tossed into public network and never reach its destination. Once I change to private VLAN interface, I no longer see the issue.

I did not know this can be such an issue. None of the baremetal kubernetes cluster setup guides mention about this.

Closing this.
